runtime/visualize.cc

```diff
@@ -0,0 +1,559 @@
+    "  }\n"
+    "  return result;\n"
+    "}\n"
+    "function slice_dim(b, d, at) {\n"
+    "  let result = clone_buffer(b);\n"
+    "  b.base += flat_offset_dim(b.dims[d], at);\n"
+    "  b.dims.splice(d, 1);\n"
+    "  return result;\n"
+    "}\n"
+    "function slice_buffer(b, at) {\n"
+    "  let result = clone_buffer(b);\n"
+    "  for (let d = at.length - 1; d >= 0; --d) {\n"
+    "    slice_dim(b, d, at[d]);\n"
+    "  }\n"
+    "  return result;\n"
+    "}\n"
+    "function truncate_rank(b, rank) {\n"
+    "  let result = clone_buffer(b);\n"
+    "  b.dims.length = rank;\n"
+    "  return result;\n"
+    "}\n"
+    "function produce(b) {\n"
+    "  for (let m of __heap_map) {\n"
+    "    if (m.begin <= b.base && b.base < m.end) {\n"
+    "      m.element.productions.push({t: __event_t++, buf: clone_buffer(b)});\n"
+    "      return;\n"
+    "    }\n"
+    "  }\n"
+    "}\n"
+    "function consume(b) {}\n";
+const char* footer =
+    "let __end_t = __event_t;\n"
+    "let __event_slider = document.getElementById('event_slider');\n"
+    "let __autoplay = true;\n"
+    "__event_slider.max = __end_t - 1;\n"
+    "document.addEventListener('keyup', event => { if (event.code === 'Space') __autoplay = !__autoplay; });\n"
+    "setInterval(function() {\n"
+    "  if (__autoplay) {\n"
+    "    __current_t = (__current_t + 1) % __end_t;\n"
+    "    __event_slider.value = __current_t;\n"
+    "  }\n"
+    "}, 100);\n"
+    "</script></body></html>\n";
+
+}  // namespace
+
+void visualize(const char* filename, const pipeline& p, pipeline::scalars args, pipeline::buffers inputs,
+    pipeline::buffers outputs, const node_context* ctx) {
+  std::ofstream file(filename);
+  file << header;
+  js_printer jsp(file, ctx);
+
+  // Print function declaration.
+  file << "function pipeline(";
+  std::vector<var> symbols = p.args();
+  symbols.insert(symbols.end(), p.inputs().begin(), p.inputs().end());
+  symbols.insert(symbols.end(), p.outputs().begin(), p.outputs().end());
+  jsp.print_vector(symbols);
+  file << ") {\n";
+
+  // Print body.
+  jsp.depth++;
+  jsp << p.body();
+  jsp.depth--;
+  file << "}\n";
+
+  // Define arguments.
+  for (index_t i = 0; i < static_cast<index_t>(p.args().size()); ++i) {
+    jsp << "let " << p.args()[i] << " = " << args[i] << ";\n";
+  }
+  for (index_t i = 0; i < static_cast<index_t>(p.inputs().size()); ++i) {
+    jsp << "let " << p.inputs()[i] << " = allocate('" << p.inputs()[i] << "', "
+        << static_cast<index_t>(inputs[i]->elem_size) << ", [";
+    jsp.print_vector(span<dim>(inputs[i]->dims, inputs[i]->rank));
+    jsp << "]);\n";
+  }
+  for (index_t i = 0; i < static_cast<index_t>(p.outputs().size()); ++i) {
+    jsp << "let " << p.outputs()[i] << " = allocate('" << p.outputs()[i] << "', "
+        << static_cast<index_t>(outputs[i]->elem_size) << ", [";
+    jsp.print_vector(span<dim>(outputs[i]->dims, outputs[i]->rank));
+    jsp << "]);\n";
+  }
+
+  // Call the pipeline.
```

Comment:
Eventually let's add a test for this, if for no other reason than it would insert a nice example of what expected JS output would look like

Reply:
I was thinking about how best to do this. I don't like the idea of just making a golden result, but I also don't know how to validate the output either.

A cool thing to do at some point would be to hook up such generated visualizations into the README somehow, and maybe at that point testing this would be more easily doable (at the very least, inspecting the readme is a lot easier than setting up a visualize use case and running it).